File: Current/Product/Production/Common/Common.Tasks/LoopThrough.vb
Option Explicit On
Option Strict On

Imports NAnt.Core
Imports NAnt.Core.Attributes


<TaskName("loopthrough")> _
Public Class LoopThrough
    Inherits Task

    Private _Items As LoopItemContainer
    Private _Actions As TaskContainer
    Private _PropertyName As String

    <TaskAttribute("property", Required:=True)> _
    Public Property PropertyName() As String
        Get
            Return _PropertyName
        End Get
        Set(ByVal value As String)
            _PropertyName = value
        End Set
    End Property

    <BuildElement("items", Required:=True)> _
    Public Property Items() As LoopItemContainer
        Get
            Return _Items
        End Get
        Set(ByVal value As LoopItemContainer)
            _Items = value
        End Set
    End Property

    <BuildElement("do", Required:=True)> _
    Public Property Actions() As TaskContainer
        Get
            Return _Actions
        End Get
        Set(ByVal value As TaskContainer)
            _Actions = value
        End Set
    End Property

    Protected Overrides Sub ExecuteTask()
        For Each Item As String In Me.Items
            If Me.Properties.Contains(Me.PropertyName) Then
                Me.Properties(Me.PropertyName) = Item
            Else
                Me.Properties.Add(Me.PropertyName, Item)
            End If
            Me.Items.Items.Executing(Item)
            Me.Actions.Execute()
        Next
    End Sub

End Class

File: Product/Production/Nant/CIFactory.NAnt.Tasks/Tasks/LoopThrough.cs
using System;
using NAnt.Core;
using NAnt.Core.Attributes;
using CIFactory.NAnt.Types;

namespace CIFactory.NAnt.Tasks
{
    [TaskName("loopthrough")]
    public class LoopThrough : Task
    {
        #region Fields

        private TaskContainer _Actions;

        private LoopItemContainer _Items;

        private string _PropertyName;

        #endregion

        #region Properties

        [BuildElement("do", Required = true)]
        public TaskContainer Actions
        {
            get { return _Actions; }
            set { _Actions = value; }
        }

        [BuildElement("items", Required = true)]
        public LoopItemContainer Items
        {
            get { return _Items; }
            set { _Items = value; }
        }

        [TaskAttribute("property", Required = true)]
        public string PropertyName
        {
            get { return _PropertyName; }
            set { _PropertyName = value; }
        }

        #endregion

        #region Protected Methods

        protected override void ExecuteTask()
        {
            foreach (string Item in this.Items)
            {
                if (this.Properties.Contains(this.PropertyName))
                {
                    this.Properties[this.PropertyName] = Item;
                }
                else
                {
                    this.Properties.Add(this.PropertyName, Item);
                }
                this.Items.Items.Executing(Item);
                this.Actions.Execute();
            }
        }

        #endregion

    }

}